File: A2Z_striver/patterns/pattern_21/pattern_21.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

// Header Comment Block
// Problem: Print the Number Spiral (Pattern 22).
// Example Input: n = 4
// Example Output:
// 4 4 4 4 4 4 4
// 4 3 3 3 3 3 4
// 4 3 2 1 2 3 4 ...
// Goal: O(N^2) Time.

class Solution
{
public:
    // ~ Approach 1: Math / Distance Formula (Your Code)
    // Analysis: Time O(N^2) | Space O(1)
    // Pros: Most efficient, no extra memory usage.
    void pattern22_math(int n)
    {
        int size = 2 * n - 1;

        for (int i = 0; i < size; i++)
        {
            for (int j = 0; j < size; j++)
            {

                // ! Logic: Value depends on min distance from any wall
                int top = i;
                int left = j;
                int bottom = (size - 1) - i; // Distance from bottom edge
                int right = (size - 1) - j;  // Distance from right edge

                // Value = n - minimum_distance
                int min_dist = min(min(top, left), min(bottom, right));
                cout << (n - min_dist) << " ";
            }
            cout << "\n";
        }
    }

    // ~ Approach 2: Matrix Layering (Simulation)
    // Analysis: Time O(N^2) | Space O(N^2)
    // Pros: Good for visualizing "Layers" or "Rings".
    void pattern22_layers(int n)
    {
        int size = 2 * n - 1;
        // Create a 2D matrix filled with 0
        vector<vector<int>> matrix(size, vector<int>(size));

        // Loop 'layer' from 0 to n-1 (0 is outer ring, n-1 is center)
        for (int layer = 0; layer < n; layer++)
        {

            // Value to fill in this layer
            int val = n - layer;

            // Fill the square boundaries for this layer
            // Start Index: layer
            // End Index: size - 1 - layer
            int start = layer;
            int end = size - 1 - layer;

            for (int i = start; i <= end; i++)
            {
                matrix[start][i] = val; // Top Row of current ring
                matrix[end][i] = val;   // Bottom Row of current ring
                matrix[i][start] = val; // Left Col of current ring
                matrix[i][end] = val;   // Right Col of current ring
            }
        }

        // Print the matrix
        for (int i = 0; i < size; i++)
        {
            for (int j = 0; j < size; j++)
            {
                cout << matrix[i][j] << " ";
            }
            cout << "\n";
        }
    }
};

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    Solution sol;
    int n = 4;

    cout << "--- Approach 1: Math Formula ---\n";
    sol.pattern22_math(n);

    cout << "\n--- Approach 2: Matrix Layers ---\n";
    sol.pattern22_layers(n);

    return 0;
}

File: A2Z_striver/patterns/pattern_21/pattern_21.py
import sys

class Solution:
    """
    Problem: Print the Number Spiral.
    Example Input: n = 4
    Example Output:
    4 4 4 4 4 4 4
    4 3 3 3 3 3 4
    4 3 2 1 2 3 4 ...
    Goal: O(N^2) Time.
    """

    # // ~ Approach 1: Math / Distance Formula (Standard)
    # // Analysis: Time O(N^2) | Space O(1)
    # // Pros: Efficient, translates well to other languages.
    def pattern22_math(self, n):
        size = 2 * n - 1
        for i in range(size):
            for j in range(size):
                
                # // 1. Calculate Distances to all 4 walls
                top = i
                left = j
                bottom = (size - 1) - i
                right = (size - 1) - j
                
                # // 2. Find Minimum Distance
                # // The value is simply: n - min_dist
                min_dist = min(top, left, bottom, right)
                print(n - min_dist, end=" ")
            print()

    # // ~ Approach 2: Pythonic List Comprehension (One-Liner)
    # // Analysis: Time O(N^2) | Space O(N^2) (Creates a list in memory)
    # // Pros: Extremely concise. Shows mastery of Python syntax.
    def pattern22_pythonic(self, n):
        size = 2 * n - 1
        
        # // Nested List Comprehension
        # // Logic: n - min(i, j, size-1-i, size-1-j) for every i, j
        grid = [
            [n - min(i, j, size - 1 - i, size - 1 - j) for j in range(size)]
            for i in range(size)
        ]
        
        # // Print the grid
        for row in grid:
            print(*row) # '*' Unpacks the list (prints elements separated by space)

    # // ~ Approach 3: Matrix Layering (Simulation)
    # // Analysis: Time O(N^2) | Space O(N^2)
    # // Pros: Visualizes the "Onion Peeling" strategy.
    def pattern22_layers(self, n):
        size = 2 * n - 1
        # Create empty grid
        matrix = [[0] * size for _ in range(size)]
        
        # Loop from outer layer (0) to inner layer (n-1)
        for layer in range(n):
            val = n - layer
            start = layer
            end = size - 1 - layer
            
            # Fill the square ring
            for i in range(start, end + 1):
                matrix[start][i] = val # Top
                matrix[end][i] = val   # Bottom
                matrix[i][start] = val # Left
                matrix[i][end] = val   # Right
                
        # Print
        for row in matrix:
            print(*row)

# Driver Code
if __name__ == "__main__":
    solver = Solution()
    n = 4
    
    print(f"--- Approach 1: Math Formula (n={n}) ---")
    solver.pattern22_math(n)
    
    print(f"\n--- Approach 2: Pythonic One-Liner (n={n}) ---")
    solver.pattern22_pythonic(n)